File: src/ui/keys.py
"""Raw-key reader for interactive Rich screens (selection, menus, scroll).

Reads single keys from /dev/tty, mapping ANSI escape sequences to symbolic
names so callers can branch on `'up' / 'down' / 'pgup'` etc.
"""
from __future__ import annotations


def open_tty_rd():
    """Return a raw-bytes file for /dev/tty or None if unavailable."""
    try:
        return open("/dev/tty", "rb", buffering=0)
    except OSError:
        return None


_CSI_NAMES = {
    "\x1b[A": "up",
    "\x1b[B": "down",
    "\x1b[C": "right",
    "\x1b[D": "left",
    "\x1b[H": "home",
    "\x1b[F": "end",
    "\x1b[5~": "pgup",
    "\x1b[6~": "pgdn",
    "\x1b[3~": "delete",
}


def _decode(seq: str) -> str:
    if seq in _CSI_NAMES:
        return _CSI_NAMES[seq]
    if seq == "\r" or seq == "\n":
        return "enter"
    if seq == " ":
        return "space"
    if seq == "\x1b":
        return "esc"
    if seq == "\x7f" or seq == "\b":
        return "backspace"
    return seq


_ESC_TIMEOUT = 0.05
_RAW_SEQ_CAP = 32


def read_key_raw(rd) -> str:
    """Read a single key, returning the raw sequence (no symbolic decoding).

    Returns the raw escape sequence (e.g. ``"\\x1b[A"``, ``"\\x1b[<0;12;5M"``)
    or single character for non-escape keys. Empty string on EOF.
    Callers needing symbolic names should use :func:`read_key`; callers needing
    raw bytes (e.g. SGR mouse parsing) should use this.
    """
    import select
    import termios
    import tty
    fd = rd.fileno()
    old = termios.tcgetattr(fd)
    try:
        tty.setcbreak(fd)
        ch = rd.read(1)
        if not ch:
            return ""
        if ch != b"\x1b":
            return ch.decode("utf-8", "ignore")
        # Disambiguate standalone Esc from start of escape sequence.
        ready, _, _ = select.select([fd], [], [], _ESC_TIMEOUT)
        if not ready:
            return "\x1b"
        b1 = rd.read(1)
        if not b1:
            return "\x1b"
        if b1 != b"[":
            return "\x1b" + b1.decode("ascii", "ignore")
        seq = b"\x1b["
        while True:
            b = rd.read(1)
            if not b:
                break
            seq += b
            # CSI terminators: any letter (incl. mouse 'M'/'m') or '~'.
            if b.isalpha() or b == b"~":
                break
            if len(seq) > _RAW_SEQ_CAP:
                break
        return seq.decode("ascii", "ignore")
    finally:
        termios.tcsetattr(fd, termios.TCSADRAIN, old)


def read_key(rd) -> str:
    """Read a single key from *rd* and return its symbolic name.

    Names: 'up','down','left','right','home','end','pgup','pgdn','delete',
    'enter','space','esc','backspace', or the literal character.
    Empty string on EOF.
    """
    raw = read_key_raw(rd)
    if raw == "":
        return ""
    return _decode(raw)

File: go/internal/keys/keys.go
// Package keys decodes raw terminal key sequences into symbolic names.
// Used by the selection TUI and plain-mode key reading.
//
// When a bubbletea TUI is active, key reading is delegated to bubbletea.
// This package provides the CSI decode table shared across TUI modes and
// a plain-mode ReadKey fallback that works with any io.Reader.
package keys

import (
	"io"
)

// CSI escape sequence → symbolic name mapping.
var csiNames = map[string]string{
	"\x1b[A":  "up",
	"\x1b[B":  "down",
	"\x1b[C":  "right",
	"\x1b[D":  "left",
	"\x1b[H":  "home",
	"\x1b[F":  "end",
	"\x1b[5~": "pgup",
	"\x1b[6~": "pgdn",
	"\x1b[3~": "delete",
}

// decode converts a raw sequence to its symbolic name.
// Returns the literal string for unrecognized sequences.
func decode(seq string) string {
	if name, ok := csiNames[seq]; ok {
		return name
	}
	switch seq {
	case "\r", "\n":
		return "enter"
	case " ":
		return "space"
	case "\x1b":
		return "esc"
	case "\x7f", "\b":
		return "backspace"
	default:
		return seq
	}
}

// ReadKeyRaw reads a single raw key sequence from rd.
// For simple io.Reader (non-TTY, or pipes): reads one byte, or attempts
// to read a CSI sequence if the first byte is Esc. Returns "" on EOF.
//
// When connected to a real TTY, the caller should use bubbletea's key
// reading or set the terminal to raw mode before calling this function.
func ReadKeyRaw(rd io.Reader) string {
	var buf [1]byte
	n, err := rd.Read(buf[:])
	if err != nil || n == 0 {
		return ""
	}
	ch := buf[0]
	if ch != 0x1b { // Esc
		return string(ch)
	}
	// Try to read CSI follow bytes.
	seq := []byte{0x1b}
	buf2 := make([]byte, 1)
	n2, err := rd.Read(buf2[:])
	if err != nil || n2 == 0 {
		return "\x1b"
	}
	if buf2[0] != '[' {
		seq = append(seq, buf2[0])
		return string(seq)
	}
	seq = append(seq, '[')
	for {
		n3, err := rd.Read(buf2[:])
		if err != nil || n3 == 0 {
			break
		}
		seq = append(seq, buf2[0])
		// CSI terminators: any letter or '~'.
		if (buf2[0] >= 'A' && buf2[0] <= 'Z') || (buf2[0] >= 'a' && buf2[0] <= 'z') || buf2[0] == '~' {
			break
		}
		if len(seq) > 32 {
			break
		}
	}
	return string(seq)
}

// ReadKey reads a single key from rd and returns its symbolic name.
// See ReadKeyRaw for raw-mode requirements on real terminals.
func ReadKey(rd io.Reader) string {
	raw := ReadKeyRaw(rd)
	if raw == "" {
		return ""
	}
	return decode(raw)
}
